pilot/pkg/networking/apigen/apigen.go

```diff
@@ -0,0 +1,165 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apigen
+
+import (
+	"strings"
+
+	gogotypes "github.com/gogo/protobuf/types"
+	golangany "github.com/golang/protobuf/ptypes/any"
+
+	mcp "istio.io/api/mcp/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/external"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/config/schema/resource"
+	"istio.io/pkg/log"
+)
+
+// APIGenerator supports generation of high-level API resources, similar with the MCP
+// protocol. This is a replacement for MCP, using XDS (and in future UDPA) as a transport.
+// Based on lessons from MCP, the protocol allows incremental updates by
+// default, using the same mechanism that EDS is using, i.e. sending only changed resources
+// in a push. Incremental deletes are sent as a resource with empty body.
+//
+// Example: networking.istio.io/v1alpha3/VirtualService
+//
+// TODO: we can also add a special marker in the header)
+type APIGenerator struct {
+}
+
+// TODO: take 'updates' into account, don't send pushes for resources that haven't changed
+// TODO: support WorkloadEntry - to generate endpoints (equivalent with EDS)
+// TODO: based on lessons from MCP, we want to send 'chunked' responses, like apiserver does.
+// A first attempt added a 'sync' record at the end. Based on feedback and common use, a
+// different approach can be used - for large responses, we can mark the last one as 'hasMore'
+// by adding a field to the envelope.
+
+// Generate implements the generate method for high level APIs, like Istio config types.
+// This provides similar functionality with MCP and :8080/debug/configz.
+//
+// Names are based on the current resource naming in istiod stores.
+func (g *APIGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates model.XdsUpdates) model.Resources {
+	resp := []*golangany.Any{}
+
+	// Note: this is the style used by MCP and its config. Pilot is using 'Group/Version/Kind' as the
+	// key, which is similar.
+	//
+	// The actual type in the Any should be a real proto - which is based on the generated package name.
+	// For example: type is for Any is 'type.googlepis.com/istio.networking.v1alpha3.EnvoyFilter
+	// We use: networking.istio.io/v1alpha3/EnvoyFilter
+	gvk := strings.SplitN(w.TypeUrl, "/", 3)
+	if len(gvk) == 3 {
+		rgvk := resource.GroupVersionKind{
+			Group:   gvk[0],
+			Version: gvk[1],
+			Kind:    gvk[2],
+		}
+		if w.TypeUrl == collections.IstioMeshV1Alpha1MeshConfig.Resource().GroupVersionKind().String() {
+			meshAny, err := gogotypes.MarshalAny(push.Mesh)
```

Comment:
I am not sure when to use golang/gogo protobuf. But it seems gogo is not actively maintained, when we bump protobuf version, we may meet some issue.

Reply:
I haven't heard about gogo not actively maintained - since K8S and Istio internally use gogo I suspect someone will pick up maintaining it. 

I don't care too much - it shouldn't matter what library you use for proto, all should be compatible and interoperable. But the purpose of this PR is not to change how K8S and Istio objects are handled - just to write code that works with both.

Reply:
Yeah, this is a confusion from me. Not related